src/passes/TypeGeneralizing.cpp

```diff
@@ -0,0 +1,438 @@
+  void visitAtomicRMW(AtomicRMW* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitAtomicCmpxchg(AtomicCmpxchg* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitAtomicWait(AtomicWait* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitAtomicNotify(AtomicNotify* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitAtomicFence(AtomicFence* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitSIMDExtract(SIMDExtract* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitSIMDReplace(SIMDReplace* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitSIMDShuffle(SIMDShuffle* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitSIMDTernary(SIMDTernary* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitSIMDShift(SIMDShift* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitSIMDLoad(SIMDLoad* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitSIMDLoadStoreLane(SIMDLoadStoreLane* curr) {
+    WASM_UNREACHABLE("TODO");
+  }
+  void visitMemoryInit(MemoryInit* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitDataDrop(DataDrop* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitMemoryCopy(MemoryCopy* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitMemoryFill(MemoryFill* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitConst(Const* curr) {}
+  void visitUnary(Unary* curr) {}
+  void visitBinary(Binary* curr) {}
+  void visitSelect(Select* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitDrop(Drop* curr) {
+    if (curr->type.isRef()) {
+      pop();
+    }
+  }
+  void visitReturn(Return* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitMemorySize(MemorySize* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitMemoryGrow(MemoryGrow* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitUnreachable(Unreachable* curr) {
+    // Require nothing about values flowing into an unreachable.
+    clearStack();
+  }
+  void visitPop(Pop* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitRefNull(RefNull* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitRefIsNull(RefIsNull* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitRefFunc(RefFunc* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitRefEq(RefEq* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitTableGet(TableGet* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitTableSet(TableSet* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitTableSize(TableSize* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitTableGrow(TableGrow* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitTableFill(TableFill* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitTry(Try* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitThrow(Throw* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitRethrow(Rethrow* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitTupleMake(TupleMake* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitTupleExtract(TupleExtract* curr) { WASM_UNREACHABLE("TODO"); }
+  void visitRefI31(RefI31* curr) { pop(); }
+  void visitI31Get(I31Get* curr) {
+    // Do not allow relaxing to nullable if the input is already non-nullable.
```

Comment:
Allowing this relaxation might result in the engine doing an extra null check, although I guess disallowing this relaxation might prevent us from removing a previous explicit null check in principle. Still, I think even when we could have removed a previous explicit null check, that's not worth making the input here nullable.

Reply:
Makes sense this might help or hurt, but isn't that the same as casts in general? Generalizing can allow removing casts - to non-null or to a heap type - but it can also hurt at runtime in theory. Is there a reason to consider nullability differently than heap types here?

Reply:
The difference is that eventually there needs to be a null check eventually, whether explicit or implicit, before doing any kind of access operation. For heap type casts, on the other hand, if we can weaken or remove a cast, that doesn't mean the cast is implicitly done later. So removing heap type casts is more useful than removing null checks.

Reply:
Fair enough. Please add a comment with that, then.

Reply:
Oh, I just remembered that engines can do virtual memory tricks for null checks as well, so the implicit null checks can actually be free. That means that removing explicit null checks is more useful than I thought, so I'll simplify this code to allow the full relaxation.

Reply:
Oh, but on the other hand virtual memory tricks for i31ref do not work because they are never dereferenced.

Reply:
I decided to go ahead and simplify this 😅